src/compiler/checker.ts

```diff
@@ -3595,13 +3600,13 @@ module ts {
             // Third, check if both types are part of deeply nested chains of generic type instantiations and if so assume the types are
             // equal and infinitely expanding. Fourth, if we have reached a depth of 100 nested comparisons, assume we have runaway recursion
             // and issue an error. Otherwise, actually compare the structure of the two types.
-            function objectTypeRelatedTo(source: ObjectType, target: ObjectType, reportErrors: boolean): Ternary {
+            function objectTypeRelatedTo(source: ObjectType, target: ObjectType, reportErrors: boolean, mustRecompute = false): Ternary {
                 if (overflow) {
                     return Ternary.False;
                 }
                 var id = relation !== identityRelation || source.id < target.id ? source.id + "," + target.id : target.id + "," + source.id;
                 var related = relation[id];
-                if (related !== undefined) {
+                if (!mustRecompute && related !== undefined) {

```

Comment:
I was thinking the same, though I'm not sure it's possible to get into that state (you would only ever be here if they weren't compatible, I think). I'm going to update the PR so that we a) don't recompute the result when it's compatible and b) don't re-elaborate.